### PR TITLE
Fix redundant this alias in loops

### LINKED ISSUE #120: Loops create redundant "this" variables
Using `this` keyword inside a loop creates redundant variables that aren't necessary because loops don't create closures (`this` inside a loop should always point to the same object as outside). 

[Minimal example](https://buble.surge.sh/#function%20bar()%20%7B%0A%20%20while%20(true)%20this%3B%0A%7D):

```js
function bar() {
  while (true) this;
}
```

Produces:

```js
function bar() {
  var this$1 = this;

  while (true) { this$1; }
}
```

Expected output:

```js
function bar() {
  while (true) { this; }
}
```

Same behavior for `for` and `do ... while` loops. Looks like a bug to me — a harmless one in practice, but quite unpleasant for perfectionists like me. :)

### PR DESCRIPTION
**Pull request by mourner**
Fixes #120. This turned out to be more involved that I thought (because loop rewriting is initialized after all its children nodes), but the logic should be more sound now.

### CODE REVIEW

**Comment by mourner** [comment]
@adrianheine any thoughts on this one?

File: src/program/types/ThisExpression.js [hunk 1]
```diff
@@ -3,18 +3,22 @@ import { loopStatement } from '../../utils/patterns.js';
 
 export default class ThisExpression extends Node {
 	initialise(transforms) {
+		const lexicalBoundary = this.findLexicalBoundary();
+
+		if (transforms.letConst) {
+			// save all loops up to the lexical boundary in case we need
+			// to alias them later for block-scoped declarations
+			let node = this.findNearest(loopStatement)
```
Comment (adrianheine) [inline src/program/types/ThisExpression.js]: Semicolon missing :)

**Comment by adrianheine** [comment]
Thanks, I will merge this if you squash the commits into one and add the missing semicolon.

**Comment by mourner** [comment]
@adrianheine thank you for the review! Updated and squashed.

**Comment by adrianheine** [comment]
Thank you for your work and patience :)

**Pull request merged**